ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11vertex.cpp

```diff
@@ -16,23 +16,279 @@
+    if (!bufferPeer) {
+      // create either a static or dynamic peer, depending on if the user called
+      // freeze on the buffer, and initialize its contents
+      D3D11_BUFFER_DESC bd;
+      bd.Usage = frozen ? D3D11_USAGE_IMMUTABLE : D3D11_USAGE_DYNAMIC;
+      bd.ByteWidth = size;
+      bd.BindFlags = isIndex ? D3D11_BIND_INDEX_BUFFER : D3D11_BIND_VERTEX_BUFFER;
+      bd.CPUAccessFlags = frozen ? 0 : D3D11_CPU_ACCESS_WRITE;
+      bd.MiscFlags = 0;
+      bd.StructureByteStride = 0;
+      D3D11_SUBRESOURCE_DATA initData;
+      initData.pSysMem = data;
+      m_device->CreateBuffer(&bd, &initData, &bufferPeer);
+
+      if (isIndex) {
+        indexBufferPeers[buffer] = bufferPeer;
+      } else {
+        vertexBufferPeers[buffer] = bufferPeer;
+      }
+    } else {
+      // update the contents of the native peer on the GPU
+      m_deviceContext->UpdateSubresource(bufferPeer, 0, NULL, data, 0, 0);
+    }
+
+    if (isIndex) {
+      indexBuffers[buffer]->indices.clear();
+      indexBuffers[buffer]->dirty = false;
+    } else {
+      vertexBuffers[buffer]->vertices.clear();
+      vertexBuffers[buffer]->dirty = false;
+    }
+  }
+}
+
+inline ID3D11InputLayout* vertex_format_layout(const enigma::VertexFormat* vertexFormat, UINT &stride) {
+  vector<D3D11_INPUT_ELEMENT_DESC> vertexLayoutElements(vertexFormat->flags.size());
+
+  size_t offset = 0;
+  map<int,int> useCount;
+  for (size_t i = 0; i < vertexFormat->flags.size(); ++i) {
+    const pair<int, int> flag = vertexFormat->flags[i];
+    D3D11_INPUT_ELEMENT_DESC *vertexLayoutElement = &vertexLayoutElements[i];
+
+    vertexLayoutElement->SemanticName = semantic_names[flag.second];
+    vertexLayoutElement->SemanticIndex = useCount[flag.second]++;
+    vertexLayoutElement->Format = dxgi_formats[flag.first];
+    vertexLayoutElement->InputSlot = 0;
+    vertexLayoutElement->AlignedByteOffset = D3D11_APPEND_ALIGNED_ELEMENT;
+    vertexLayoutElement->InputSlotClass = D3D11_INPUT_PER_VERTEX_DATA;
+    vertexLayoutElement->InstanceDataStepRate = 0;
 
-void graphics_delete_index_buffer_peer(int buffer) {}
+    offset += dxgi_format_sizes[flag.first];
+  }
+  stride = offset;
+
+  ID3D11InputLayout* vertexLayout;
+  m_device->CreateInputLayout(vertexLayoutElements.data(), vertexLayoutElements.size(), pBlobVS->GetBufferPointer(),
+                              pBlobVS->GetBufferSize(), &vertexLayout);
+
+  return vertexLayout;
+}
+
+void graphics_prepare_default_shader() {
+  static ID3D11VertexShader* g_pVertexShader = NULL;
+  static ID3D11PixelShader* g_pPixelShader = NULL;
+
+  if (g_pVertexShader == NULL) {
+    DWORD dwShaderFlags = D3D10_SHADER_ENABLE_STRICTNESS;
+#ifdef DEBUG_MODE
+    dwShaderFlags |= D3D10_SHADER_DEBUG | D3D10_SHADER_SKIP_OPTIMIZATION;
+#endif
+
+    ID3D10Blob* pBlobError = NULL;
+    HRESULT hr;
+
+    // create the vertex shader
+    hr = D3DCompile(g_strVS, lstrlenA(g_strVS) + 1, "VS", NULL, NULL, "VS",
```

Comment:
This block differs from the block below by two constants and two variables.

Reply:
The better news is that there's no second copy of the code to maintain. I sort of expected you to make the final function call in there, too, though, using a boolean. Regardless, it's fine the way it is.

Reply:
Sure, I am not worried about it because all of this code is going to get moved out of here anyway when I do the full HLSL11 shader support. Don't worry about it for now I am happy with it.